checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+                CheckerError::Mumble("Failed to register user")
+            })?;
+        self.conn.flush().await.map_err(|e| {
+            info!("Flush failed -- {}", e);
+            CheckerError::Mumble("Falied to register user")
+        })?;
+
+        debug!("Waiting for registration to complete");
+        let mut response = Vec::with_capacity(100);
+        match read_until_slice(&mut self.conn, PROMPT, &mut response).await {
+            Err(e) => {
+                info!("Socket error waiting for user registration {:?}", e);
+                Err(CheckerError::Mumble("Registration failed"))
+            }
+            Ok(0) => {
+                info!("Unexpected EOF waiting for user registration");
+                Err(CheckerError::Mumble("Registration failed"))
+            }
+            Ok(r) => Ok(r),
+        }?;
+
+        if response != b"User successfully registered\n> " {
+            info!("Unexpected response: {}", bytes_debug_repr(&response));
+            return Err(CheckerError::Mumble("Registration Failed"));
+        }
+
+        Ok(())
+    }
+
+    #[instrument("LOGIN")]
+    pub async fn login(&mut self, user: NotebookUser) -> CheckerResult<()> {
+        info!("Logging in as {:?}", &user);
+        self.user = Some(user);
+        let user_ref = self.user.as_ref().unwrap();
+        let conn = &mut self.conn;
+
+        let mut response_buf = Vec::with_capacity(64);
+        async {
+            let login_str = format!("log {} {}\n", user_ref.username, user_ref.password);
+            conn.write_all(login_str.as_bytes()).await?;
+            conn.flush().await?;
+
+            debug!("Waiting for login");
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e: std::io::Error| {
+            warn!("Connection-Error on Login: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Login failed")
+        })?;
+
+        if response_buf != b"Successfully logged in!\n> " {
+            info!("Unexpected response: {}", bytes_debug_repr(&response_buf));
+            return Err(CheckerError::Mumble("Login failed"));
+        }
+
+        Ok(())
+    }
+
+    #[instrument("SET_NOTE")]
+    pub async fn set_note(&mut self, note: &str) -> CheckerResult<()> {
+        info!("Deploying Note {}", note);
+
+        let conn = &mut self.conn;
+
+        let mut response_buf = Vec::with_capacity(64);
+        async {
+            let request = format!("set {}\n", note);
+            debug!("running: '{}'", request);
+            conn.write_all(request.as_bytes()).await?;
+            conn.flush().await?;
+
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e| {
+            warn!("Connection-Error: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Set-Note connection error")
+        })?;
+
+        let response = String::from_utf8(response_buf)?;
```

Comment:
Will lead to `InternalError` if response contains illegal uf8!

Reply:
impl `From<FromUTF8Error>` for `CheckerError` returns Mumble

Reply:
Not a good solution:
We had a long discussion about this for the python lib and finally all agreed that this is a bad idea, see https://github.com/enowars/enochecker/issues/77

There are many possible fields in the service where other teams may deposit broken utf8, rendering a service mumble - in an upatchable way - while the real reason is that the checker authors are too stupid to write their code correctly.
We should either have a "safe" version in the library (as implemented above), or let the authors handle it explicitly.

Reply:
To quote @MMunier on this issue: "I'd disagree that the checker should handle that, since in most cases UnicodeDecodeErrors mean that the checker-script is kind of broken. Seeing a random checker error appear is exactly the right response to that."

Reply:
The difference being, here you have to explicitly handle the Error ( by writing ? ) whereas in Python all errors of this kind are caught under the same umbrella if we use a big try-catch.
Not the biggest of differences but enough to warrant Mumble as the default.

Reply:
I don't know if thats a huge difference, but fine for now, we should probably move the discussion to https://github.com/enowars/enochecker-rs/issues/2